sdk/core/azure-core/src/main/java/com/azure/core/util/paging/ContinuablePagedFluxCore.java

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.paging;
+
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * The default implementation of {@link ContinuablePagedFlux}.
+ *
+ * The Flux that provides the ability to operate on individual items in pages of type {@link ContinuablePage},
+ * it also provide ability to operate on individual pages.
+ *
+ * The constructor of this type takes a provider method, that when called should provides Page Retriever
+ * Function which accepts continuation token. The provider is called for each Subscription to the
+ * ContinuablePagedFluxCore instance. Given provider is called per Subscription, the provider implementation
+ * can create one or more objects to store any state and Page Retriever Function can capture and use those objects.
+ * This indirectly associate the state objects to the Subscription. The Page Retriever Function can get called
+ * multiple times in serial fashion, each time after the completion of the Flux returned by the previous invocation.
+ * The final completion signal will be send to the downstream subscriber when the last Page emitted by the Flux
+ * returned by Page Continuation Function has {@code null} continuation token.
+ *
+ * <p><strong>Extending PagedFluxCore for Custom Continuation Token support</strong></p>
+ * {@codesnippet com.azure.core.util.paging.pagedfluxcore.continuationtoken}
+ *
+ * @param <T> The type of items in a {@link ContinuablePage}
+ * @param <P> The {@link ContinuablePage} holding items of type {@code T}.
+ *
+ * @see ContinuablePage
+ * @see Flux
+ */
+public abstract class ContinuablePagedFluxCore<C, T, P extends ContinuablePage<C, T>>
+    extends ContinuablePagedFlux<C, C, T, P> {
+
+    private final Supplier<Function<C, Flux<P>>> pageRetrieverProvider;
+    private final int defaultPrefetch;
+
+    /**
+     * Creates an instance of {@link ContinuablePagedFluxCore}.
+     *
+     * @param pageRetrieverProvider a provider that returns Page Retriever Function.
+     */
+    protected ContinuablePagedFluxCore(Supplier<Function<C, Flux<P>>> pageRetrieverProvider) {
```

Comment:
It may be all the generic type arguments that are throwing me off, but why provide a supplier that returns a function rather than a reference to the function itself? (I was trying to see a concrete use case in the code snippets/samples but couldn't see when it would change.)

Reply:
This is to support optional state-management. 90% cases the string continuation-token is the state that flows between Page retrieval Function calls.  hence it will be

```java
() -> {
   return new Function<String, Flux<Page<Foo>>() {
     // page retrieval api call
   }
} 
```

When exploring Cosmos pagination we learned that some client SDK might need to manage internal state that is accessible between multiple call to Page Retrieval Function. Below is a made up example showcasing this use case.

```java
() -> {
    State state = new State();

    return new Function<String, Flux<Page<Foo>>() {
         if (firstCall) {
           result = make api call
           state.store(result.metadata)
           return Flux.just(new Page(result.items, result.link));
        } else {
           metadata = state.get();
           // use the metadata to prepare arguments for next call
           // result  = make api call
           return Flux.just(new Page(result.items, result.link));
        }
   }
}
```

This Supplier will be called for each Subscription to the Paged Flux instance, which means state per Subscription, the Page retrieval Function instance will capture this state, so throughout the life time of a Subscription the Page retrieval Function instance uses the same state.